### PR TITLE
Update openai autolog

### PR DESCRIPTION
**Pull request by harupy**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/harupy/mlflow/pull/13248?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/13248/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 13248
```

</p>
</details>

### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?

title

### How is this PR tested?

- [x] Existing unit/integration tests
- [ ] New unit/integration tests
- [x] Manual tests

```python
import os

import openai

import mlflow

assert "OPENAI_API_KEY" in os.environ, "Please set the OPENAI_API_KEY environment variable."

mlflow.openai.autolog(log_models=True)

messages = [
    {
        "role": "user",
        "content": "tell me a joke in 50 words",
    }
]

output = openai.chat.completions.create(
    model="gpt-4o-mini",
    messages=messages,
    temperature=0,
)

print(mlflow.search_logged_models())
```

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [x] No (this PR will be included in the next minor release)

### CODE REVIEW

File: mlflow/openai/_openai_autolog.py [hunk 1]
```diff
@@ -200,9 +200,9 @@ def _stream_output_logging_hook(stream: Iterator) -> Iterator:
                 # so that the model can be logged.
                 with _set_api_key_env_var(self._client):
                     mlflow.openai.log_model(
-                        kwargs.get("model", None),
-                        task,
-                        "model",
+                        model=kwargs.get("model", None),
```
Comment (dbczumar) [inline mlflow/openai/_openai_autolog.py:203]: Can we update this to link the traces to the model as well? See the demo notebook for an example.

Comment (harupy) [reply mlflow/openai/_openai_autolog.py:203]: got it, will check the notebook

Comment (dbczumar) [reply mlflow/openai/_openai_autolog.py:203]: https://github.com/mlflow/mlflow/blob/a049494a66bee10267435566c01a4b6cb0a2bc15/mlflow/tracing/fluent.py#L192

File: mlflow/openai/_openai_autolog.py [hunk 2]
```diff
@@ -120,9 +120,47 @@ def patched_call(original, self, *args, **kwargs):
...
+                    model_id = logged_model.model_id
+        except Exception as e:
+            _logger.warning(f"Failed to log model due to error: {e}")
+        self._mlflow_model_logged = True
+
     if config.log_traces:
         root_span = mlflow_client.start_trace(
-            name=self.__class__.__name__, span_type=_get_span_type(self.__class__), inputs=kwargs
+            name=self.__class__.__name__,
+            span_type=_get_span_type(self.__class__),
```
Comment (dbczumar) [inline mlflow/openai/_openai_autolog.py:162]: If I do `mlflow.pyfunc.load_model()` or `mlflow.openai.load_model()` can call the model again with autologging enabled, can I link the trace to the ID of the model that I just loaded?

Comment (harupy) [reply mlflow/openai/_openai_autolog.py:162]: That's possible. This is a dumb approach but should work:

```python
_LOADED_MODEL_ID = None  # set this in load_model

# in autolog

if _LOADED_MODEL_ID:
    # skip model logging
    
    # log trace with model=_LOADED_MODEL_ID
```

Comment (harupy) [reply mlflow/openai/_openai_autolog.py:162]: what should happen if you create a new OpenAI model and chat with it?

Comment (harupy) [reply mlflow/openai/_openai_autolog.py:162]: Can I address this in a separate PR?

Comment (dbczumar) [reply mlflow/openai/_openai_autolog.py:162]: If I create a new model, it should be logged as a new model

Comment (dbczumar) [reply mlflow/openai/_openai_autolog.py:162]: Yeah, a separate PR is fine - let's make sure to file a ticket or project plan entry to track it

File: mlflow/openai/_openai_autolog.py [hunk 3]
```diff
@@ -120,9 +120,47 @@ def patched_call(original, self, *args, **kwargs):
...
+            attributes={"model_id": model_id} if model_id else None,
```
Comment (dbczumar) [inline mlflow/openai/_openai_autolog.py:163]: @harupy Is this the right attribute name? I'm pretty sure we should be using `SpanAttributeKey.MODEL_ID`

**Review by dbczumar — APPROVED** [review]
LGTM once the model ID attribute has been renamed

File: mlflow/openai/_openai_autolog.py (hunk 3 above)
Comment (harupy) [reply mlflow/openai/_openai_autolog.py:163]: didn't Know SpanAttributeKey.MODEL_ID exists, will use it

File: mlflow/openai/_openai_autolog.py (hunk 2 above)
Comment (harupy) [reply mlflow/openai/_openai_autolog.py:162]: added an entry to the project plan.

File: mlflow/openai/_openai_autolog.py (hunk 3 above)
Comment (harupy) [reply mlflow/openai/_openai_autolog.py]: updated

**Pull request merged**